### PR TITLE
refactor chunked serialization 

### PR DESCRIPTION
**Pull request by vinniefalco**
* buffers_ref is Assignable

HTTP

* Shrink chunk header buffer sequence size

API Changes:

* Refactor chunked-encoding serialization

Actions Required:

* Remove references to ChunkDecorators. Use the new chunk-encoding
  buffer sequences to manually produce a chunked payload body in
  the case where control over the chunk-extensions and/or trailers
  is required.

Documentation preview:
http://vinniefalco.github.io/stage/beast/bjorn/beast/using_http/chunked_encoding.html

### CODE REVIEW

File: doc/qbk/04_http/08_chunked_encoding.qbk [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+[section Chunked Encoding]
+
+For message payloads whose size is not known ahead of time, HTTP
+version 1.1 defines the
+[@https://tools.ietf.org/html/rfc7230#section-4.1 ['chunked]]
+transfer coding. This coding consists of zero or more
+[@https://tools.ietf.org/html/rfc7230#section-4.1 ['chunked bodies]],
+followed by a
+[@https://tools.ietf.org/html/rfc7230#section-4.1 ['last chunk]].
+Each chunked body may contain optional appliation-defined, connection-specific
```
Comment (mpark) [inline doc/qbk/04_http/08_chunked_encoding.qbk]: `s/appliation/application/`

File: doc/qbk/04_http/08_chunked_encoding.qbk [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+    [
+        A buffer sequence representing a last chunk. The last chunk
+        indicates the end of the chunked message payload, and may
+        contain optional trailer fields.
+    ]
+][
+    [
+        [link beast.ref.beast__http__make_chunk `make_chunk`]
+
+        [link beast.ref.beast__http__make_chunk `make_chunk_last`]
```
Comment (mpark) [inline doc/qbk/04_http/08_chunked_encoding.qbk]: Is this supposed to point to `beast.ref.beast__http__make_chunk`?
I do see [this page](http://vinniefalco.github.io/stage/beast/bjorn/beast/ref/beast__http__make_chunk_last.html) but not sure if this was a mistake.

Comment (vinniefalco) [inline doc/qbk/04_http/08_chunked_encoding.qbk]: Typo! willfix

File: doc/qbk/04_http/08_chunked_encoding.qbk (hunk 1 above)
Comment (vinniefalco) [inline doc/qbk/04_http/08_chunked_encoding.qbk]: willfix

File: include/beast/core/detail/buffers_ref.hpp [hunk 3]
```diff
@@ -27,23 +27,24 @@ class buffers_ref
         typename BufferSequence::const_iterator;
 
     buffers_ref(buffers_ref const&) = default;
+    buffers_ref& operator=(buffers_ref const&) = default;
 
     explicit
     buffers_ref(BufferSequence const& buffers)
-        : buffers_(buffers)
+        : buffers_(&buffers)
```
Comment (glenfe) [inline include/beast/core/detail/buffers_ref.hpp]: boost::addressof(buffers) instead of &buffers

Comment (vinniefalco) [inline include/beast/core/detail/buffers_ref.hpp]: Some day I will remember this on my own. Not today though.

File: include/beast/http/detail/chunk_encode.hpp [hunk 4]
```diff
@@ -17,73 +19,109 @@ namespace beast {
...
+        : ext_(std::move(ext))
+    {
+    }
+
+    chunk_extensions_impl(ChunkExtensions const& ext)
+        : ext_(ext)
+    {
+    }
+
+    virtual
```
Comment (glenfe) [inline include/beast/http/detail/chunk_encode.hpp]: No need to repeat 'virtual' here; 'override' has you covered.

File: include/beast/http/detail/chunk_encode.hpp [hunk 5]
```diff
@@ -17,73 +19,109 @@ namespace beast {
...
+    }
+};
+
+template<class T, class = void>
+struct is_chunk_extensions : std::false_type {};
+
+template<class T>
+struct is_chunk_extensions<T, beast::detail::void_t<decltype(
+    std::declval<string_view&>() = std::declval<T&>().str(),
+        (void)0)>> : std::true_type
```
Comment (glenfe) [inline include/beast/http/detail/chunk_encode.hpp]: `void_t<expression,(void)0>` not necessary here, is it? Just:  `void_t<expression>` should suffice.

File: include/beast/http/detail/chunk_encode.hpp (hunk 4 above)
Comment (vinniefalco) [inline include/beast/http/detail/chunk_encode.hpp]: copypasta willfix

File: include/beast/http/detail/chunk_encode.hpp (hunk 5 above)
Comment (vinniefalco) [inline include/beast/http/detail/chunk_encode.hpp]: idk, I tend to just copy and paste these things without understanding them. I might copy this one and if I remove the `(void)0` you will hear me asking questions on Slack

**Comment by vinniefalco** [comment]
Everything has been addressed in https://github.com/vinniefalco/Beast/pull/641 thanks!

**Pull request closed without merging**